src/coprocessor/dag/rpn_expr/impl_cast.rs

```diff
@@ -0,0 +1,203 @@
+// Copyright 2019 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::convert::TryFrom;
+
+use cop_codegen::RpnFunction;
+use cop_datatype::{EvalType, FieldTypeAccessor, FieldTypeFlag};
+use tipb::expression::FieldType;
+
+use crate::coprocessor::codec::data_type::*;
+use crate::coprocessor::dag::expr::EvalContext;
+use crate::coprocessor::dag::rpn_expr::function::RpnFunction;
+use crate::coprocessor::dag::rpn_expr::types::RpnFnCallPayload;
+use crate::coprocessor::Result;
+
+/// Gets the cast function between specified data types.
+///
+/// TODO: This function supports some internal casts performed by TiKV. However it would be better
+/// to be done in TiDB.
+pub fn get_cast_fn(
+    from_field_type: &FieldType,
+    to_field_type: &FieldType,
+) -> Result<Box<dyn RpnFunction>> {
+    let from = box_try!(EvalType::try_from(from_field_type.tp()));
+    let to = box_try!(EvalType::try_from(to_field_type.tp()));
+    Ok(match (from, to) {
+        (EvalType::Int, EvalType::Decimal) => {
+            if !from_field_type
+                .as_accessor()
+                .flag()
+                .contains(FieldTypeFlag::UNSIGNED)
+                && !to_field_type
+                    .as_accessor()
+                    .flag()
+                    .contains(FieldTypeFlag::UNSIGNED)
```

Comment:
I feel strange about it...Converting a signed int in an unsigned way can lead to weird result....
I don't know how to explain the case below. I am skeptical if such a case when converting to an unsigned decimal exists.
```
create table t(col1 decimal(20, 10) unsigned, col2 bigint);
insert into t values (1, -5);

MySQL [test]> select * from t;
+--------------+------+
| col1         | col2 |
+--------------+------+
| 1.0000000000 |   -5 |
+--------------+------+
1 row in set (0.002 sec)

MySQL [test]> explain select col1 + col2 from t;
+---------------------+----------+------+------------------------------------------------------------+
| id                  | count    | task | operator info                                              |
+---------------------+----------+------+------------------------------------------------------------+
| Projection_3        | 10000.00 | root | plus(test.t.col1, cast(test.t.col2))                       |
| └─TableReader_5     | 10000.00 | root | data:TableScan_4                                           |
|   └─TableScan_4     | 10000.00 | cop  | table:t, range:[-inf,+inf], keep order:false, stats:pseudo |
+---------------------+----------+------+------------------------------------------------------------+
3 rows in set (0.001 sec)

MySQL [test]> select col1 + col2 from t;
+---------------+
| col1 + col2   |
+---------------+
| -4.0000000000 |
+---------------+
1 row in set (0.002 sec)
```

Reply:
The logic comes from TiDB:

```golang
func (b *builtinCastIntAsDecimalSig) evalDecimal(row chunk.Row) (res *types.MyDecimal, isNull bool, err error) {
	val, isNull, err := b.args[0].EvalInt(b.ctx, row)
	if isNull || err != nil {
		return res, isNull, err
	}
	if !mysql.HasUnsignedFlag(b.tp.Flag) && !mysql.HasUnsignedFlag(b.args[0].GetType().Flag) {
		res = types.NewDecFromInt(val)
	} else if b.inUnion && val < 0 {
		res = &types.MyDecimal{}
	} else {
		var uVal uint64
		sc := b.ctx.GetSessionVars().StmtCtx
		uVal, err = types.ConvertIntToUint(sc, val, types.IntergerUnsignedUpperBound(mysql.TypeLonglong), mysql.TypeLonglong)
		if err != nil {
			return res, false, err
		}
		res = types.NewDecFromUint(uVal)
	}
	res, err = types.ProduceDecWithSpecifiedTp(res, b.tp, b.ctx.GetSessionVars().StmtCtx)
	return res, isNull, err
}
```

Since `inUnion` is not supported, that branch is omitted. And now there are two branches:

Branch 1: Evaluate a signed decimal from a signed int: NewDecFromInt
Branch 2: Else: NewDecFromUint

Reply:
BTW, "Converting a signed int in an unsigned way can lead to weird result...." is not actually what happened. We are using `i64` for storing both signed and unsigned values. When storing unsigned values, they are stored by force casting via `as i64` in https://github.com/tikv/tikv/blob/0c57e3c72ef1a0808235a152f350fe709585b87e/src/coprocessor/codec/data_type/vector.rs#L291

So now we will simply cast it back to u64 when we know that it's unsigned.

Reply:
I mean the case when the int is signed but we are going to convert it into an unsigned decimal.

Reply:
I see the implementation in TiDB will generate an overflow error. But here we cast it to u64 even if it's a negative value.

Reply:
The decimal doesn't care about it is signed or not actually. The branch "else" covers:
1. unsigned decimal, signed int
2. unsigned decimal, unsigned int
3. signed decimal, unsigned int

And more precisely it *wants* to cover point 3 instead of 1 and 2.

Reply:
I understand decimal doesn't care about if it is signed or not. If I don't misunderstand you, we are going to align with TiDB, right? The problem here is we are actually not. If the int is negative and the decimal *accidentally* contains an unsigned flag, TiDB generates an error while we outputs unexpected result.

Reply:
Ah, sorry. I just find TiDB actually never raises an error here. The only case we differ is on insert or update. That has nothing to do with coprocessor. I'll mark it as resolved.

Reply:
I will introduce the `isUnion` flag to coprocess after reviewing all difference in `CAST` function implementations. Most of `CAST` functions has some corner cases different with the TiDB in the current implementation.